client/src/main/scala/com.gu.contentapi.client/BackoffStrategy.scala

```diff
@@ -85,17 +76,35 @@ object ContentApiBackoff {
   private def constant(
     min: Duration = Duration(defaultMinimumInterval, TimeUnit.MILLISECONDS),
     maxAttempts: Int = defaultMaxAttempts
-  )(implicit executor: ScheduledExecutor): Constant = {
+  ): Constant = {
     val ln = Math.max(min.toMillis, defaultMinimumInterval)
     val mx = if (maxAttempts > 0) maxAttempts else 1
     Constant(Duration(ln, TimeUnit.MILLISECONDS), 0, mx)
   }
+}
 
-  private def attempt(backoff: ContentApiBackoff, retry: Int, operation: Int => Future[HttpResponse])(implicit context: ExecutionContext): Future[HttpResponse] = {
-    operation(retry)
-      .recoverWith {
-        case _: ContentApiRecoverableException => backoff.execute(operation)
-      }
+object Retry {
+  type RetryAttempt = Int
+  def withRetry[A](backoffStrategy: BackoffStrategy, retryPredicate: A => Boolean)(operation: RetryAttempt => Future[A])(implicit executor: ScheduledExecutor, ec: ExecutionContext): Future[A] = {
```

Comment:
Generic retry which does not care about Http

Reply:
Just to check, has this been created to allow for testing retry behaviour isolated from (complexity of) HTTP domain?

Reply:
Not actually. Generally, I try to create higher-level abstractions to create simple things and avoid mixing concerns. retry-backoff has got nothing to do with HTTP or Capi content client. These concerns should not have mixed together in first place.

Reply:
I think both approaches have their disadvantages; I've definitely been guilty of trying to create higher-level abstractions, where a concrete implementation would have been simpler. Personally, I'm of the view that code should be minimal and you should only generalise when required, though I appreciate this subject can be contentious.

Reply:
For this, we shall talk outside this PR :)